addons/bus/static/src/multi_tab_service.js

```diff
@@ -0,0 +1,211 @@
+                const cleanedTabs = {};
+                for (const [tabId, lastPresence] of Object.entries(lastPresenceByTab)) {
+                    if (lastPresence + HEARTBEAT_KILL_OLD_PERIOD > now) {
+                        cleanedTabs[tabId] = lastPresence;
+                    }
+                }
+                if (heartbeatValue !== lastHeartbeat) {
+                    // Someone else is also main...
+                    // It should not happen, except in some race condition situation.
+                    _isOnMainTab = false;
+                    lastHeartbeat = 0;
+                    lastPresenceByTab[tabId] = now;
+                    setItemInStorage('lastPresenceByTab', lastPresenceByTab);
+                    bus.trigger('no_longer_main_tab');
+                } else {
+                    lastHeartbeat = now;
+                    setItemInStorage('heartbeat', now);
+                    setItemInStorage('lastPresenceByTab', cleanedTabs);
+                }
+            } else {
+                // Update own heartbeat.
+                lastPresenceByTab[tabId] = now;
+                setItemInStorage('lastPresenceByTab', lastPresenceByTab);
+            }
+            const hbPeriod = _isOnMainTab ? MAIN_TAB_HEARTBEAT_PERIOD : TAB_HEARTBEAT_PERIOD;
+            heartbeatTimeout = browser.setTimeout(heartbeat, hbPeriod);
+        }
+
+        function onStorage({ key, newValue }) {
+            if (key === generateLocalStorageKey('main') && !newValue) {
+                // Main was unloaded.
+                startElection();
+            }
+            if (PRIVATE_LOCAL_STORAGE_KEYS.includes(key)) {
+                return;
+            }
+            if (key && key.includes(localStoragePrefix)) {
+                // Only trigger the shared_value_updated event if the key is
+                // related to this service/origin.
+                const baseKey = key.replace(localStoragePrefix, '');
+                bus.trigger('shared_value_updated', { key: baseKey, newValue });
+            }
+        }
+
+        function onUnload() {
+            clearTimeout(heartbeatTimeout);
+            const lastPresenceByTab = getItemFromStorage('lastPresenceByTab', {});
+            delete lastPresenceByTab[tabId];
+            setItemInStorage('lastPresenceByTab', lastPresenceByTab);
+
+            // Unload main.
+            if (_isOnMainTab) {
+                browser.localStorage.removeItem(generateLocalStorageKey('main'));
+            }
+        }
+
+        browser.addEventListener('unload', onUnload);
+        browser.addEventListener('storage', onStorage);
+
+        // REGISTER THIS TAB
+        const lastPresenceByTab = getItemFromStorage('lastPresenceByTab', {});
+        lastPresenceByTab[tabId] = now;
+        setItemInStorage('lastPresenceByTab', lastPresenceByTab);
+
+        if (!getItemFromStorage('main')) {
+            startElection();
+        }
+        heartbeat();
+
+        return Object.assign(bus, {
```

Comment:
I think it would be cleaner to simply export an object with a bus in it:
```js
return {
  bus,
  isOnMainTab ...,
  ...
};
```
(not a huge dealbreaker, but would be safer)